src/decode.rs

```diff
@@ -0,0 +1,55 @@
+// Copyright 2019 King's College London.
+// Created by the Software Development Team <http://soft-dev.org/>.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use crate::{MetaData, SER_VERSION};
+use fallible_iterator::FallibleIterator;
+use rmp_serde::{
+    decode::{self, ReadReader},
+    Deserializer,
+};
+use serde::Deserialize;
+use std::io::Read;
+
+/// The meta-data decoder.
+/// Offfers a simple iterator interface to serialised meta-data.
+pub struct Decoder<'a> {
+    deser: Deserializer<ReadReader<&'a mut dyn Read>>,
+}
+
+impl<'a> Decoder<'a> {
+    /// Returns a new decoder which will deserialise from `read_from`.
+    pub fn from(read_from: &'a mut dyn Read) -> Result<Self, decode::Error> {
+        let mut deser = Deserializer::new(read_from);
+        let ver = usize::deserialize(&mut deser)?;
+        if ver != SER_VERSION {
+            panic!("incorrect version number!");
+        }
+        Ok(Self { deser })
+    }
+
+    // Iterate over meta-data.
+    pub fn iter(self) -> MetaDataIterator<'a> {
```

Comment:
Is it intentional that `iter` consumes `self`?

Reply:
Yeah, but if you think it's important to be able to allow the user to iterate multiple times using the same decoder instance, i can try `&mut self`. It's not immediately clear if it's possible though... I think you'd need >1 mutable ref too allow that.

Reply:
It might be better off as-is then. It might be worth documenting this, something like "Note that iterating over this structure consumes it"?

Reply:
Let me try to confirm my suspicion regarding `&muts `and get back to you.

Reply:
I've tried a few different approaches. I don't think it's possible to allow multiple iterations over the same decoder.

In light of that, I do wonder if instead of having a decoder struct from which we can get an iterator, we could have a decoder which *is* an iterator. What would you prefer?

Reply:
Good question. If it can only be a one-off iterator, then making the thole thing a one-off iterator does seem like a good idea.

Reply:
Let's try it in a separate commit, and we can revert if we hate it.

Reply:
OK.